_posts/2023-07-24-native-startup-rss-troubleshooting.adoc

```diff
@@ -0,0 +1,228 @@
+and that was specific to the Quarkus code.
+
+Following
+https://www.brendangregg.com/FlameGraphs/memoryflamegraphs.html[Brendan Gregg’s Memory Leak (and Growth) Flame Graphs guide] on a Linux environment,
+we were able to narrow down the issue to the shape of the flamegraphs for the `mmap` system calls.
+To obtain the flamegraphs, we capture the `sys_enter_mmap` system call events with `perf record`,
+and then we generate flamegraphs with the recorded data:
+
+[source,bash]
+----
+$ sudo perf record -e syscalls:sys_enter_mmap --call-graph dwarf -- getting-started/target/getting-started-1.0.0-SNAPSHOT-runner -Xmx64m
+...
+2023-07-21 10:13:11,304 INFO  [io.quarkus] (main) getting-started 1.0.0-SNAPSHOT native (powered by Quarkus 2.7.6.Final) started in 0.023s. Listening on: http://0.0.0.0:8080
+
+$ perf script -i perf.data > out.stacks
+$ /opt/FlameGraph/stackcollapse-perf.pl < out.stacks | /opt/FlameGraph/flamegraph.pl \
+   --color=mem \
+   --title="mmap Flame Graph" \
+   --countname="calls" > out.svg
+----
+
+[NOTE]
+====
+The native executable above was created passing in
+`-Dquarkus.native.debug.enabled` and `-Dquarkus.native.additional-build-args=-H:-DeleteLocalSymbols`
+as extra build flags.
+====
+
+The areas of interest in the flamegraphs are those stacktraces that contain `ApplicationImpl.doStart()` method call.
+If you focus the flamegraph on that method,
+and then click on `Search` at the top right and type `allocate`,
+you can observe which of the `sys_enter_mmap` system calls are related to native heap memory allocation.
+These allocations are for heap chunks, each of which is 1MB,
+which are part of the thread local allocation logic in GraalVM’s runtime engine, SubstrateVM
+(further details in the
+https://quarkus.io/guides/native-reference#rss[Quarkus native reference guide]
+).
+
+For Quarkus 2.7.6.Final, the flamegraph looks like this:
+
+image::rss-before-mmap.svg[startup rss startup flamegraph for Quarkus 2.7.6.Final]
+
+And here's a screenshot of the mmap calls that relate to heap chunk allocations:
+
+image::rss-before-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.7.6.Final]
+
+Observing the number of times `allocate` is highlighted above, i.e. 2, we can say that the `main` thread in Quarkus 2.7.6.Final allocates roughly 2MB of heap chunks.
+
+For Quarkus 2.13.4.Final, the flamegraph looks like this:
+
+image::rss-after-mmap.svg[startup rss startup flamegraph for Quarkus 2.13.4.Final]
+
+Focusing on the same area, we observe that the `allocate` count is 5 in this case, so the `main` thread allocates roughly 5MB of heap chunks:
+
+image::rss-after-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.13.4.Final]
+
+The increase in object allocation at runtime startup between these two Quarkus versions,
+combined with the idiosyncrasies of how thread local allocation works in SubstrateVM,
```

Comment:
Maybe it's worth elaborating on "idiosyncrasies". For example, explaining that Hotspot has dynamic TLAB resizing whereas SVM has a one-size-tries-to-fit-all approach, which is why the issue only pops up in native mode even though both modes experience more allocations.

Reply:
This has been solved in the latest update. In expanding https://github.com/quarkusio/quarkusio.github.io/pull/1755#discussion_r1272268418, I've explained how the slow/fast path work and the TLA logic in SubstrateVM. I've then expanded this paragraph further.